File: src/theme/themes.ts
/**
 * Theme definitions for the Git Changes Monitor TUI
 * Supports: Catppuccin Mocha, Nord, Tokyo Night, Dracula
 */

export type ThemeName = "catppuccin" | "nord" | "tokyo-night" | "dracula";

export interface Theme {
  name: ThemeName;
  displayName: string;

  // Base colors
  base: string; // Main background
  surface: string; // Section backgrounds (slightly lighter)
  overlay: string; // Modal/popup backgrounds

  // Text colors
  text: string; // Primary text
  subtext: string; // Dimmed/secondary text

  // Accent colors
  accent: string; // Primary accent (headers, highlights)
  accentSecondary: string; // Secondary accent

  // Semantic colors
  success: string; // Added, success states
  error: string; // Deleted, error states
  warning: string; // Modified, warnings
  info: string; // Info, renamed, copied

  // Git-specific
  staged: string; // Staged indicator
  unstaged: string; // Unstaged indicator
  untracked: string; // Untracked files

  // Selection
  selection: string; // Selected item background
  selectionText: string; // Selected item text

  // Borders (for help menu)
  border: string;
}

/**
 * Catppuccin Mocha Theme
 * Warm, pastel colors with purple/pink accents
 * https://github.com/catppuccin/catppuccin
 */
export const catppuccin: Theme = {
  name: "catppuccin",
  displayName: "Catppuccin Mocha",

  // Base colors
  base: "#1e1e2e",
  surface: "#313244",
  overlay: "#45475a",

  // Text colors
  text: "#cdd6f4",
  subtext: "#a6adc8",

  // Accent colors
  accent: "#cba6f7", // Mauve
  accentSecondary: "#f5c2e7", // Pink

  // Semantic colors
  success: "#a6e3a1", // Green
  error: "#f38ba8", // Red
  warning: "#f9e2af", // Yellow
  info: "#89b4fa", // Blue

  // Git-specific
  staged: "#94e2d5", // Teal
  unstaged: "#fab387", // Peach
  untracked: "#6c7086", // Overlay0

  // Selection
  selection: "#45475a",
  selectionText: "#cdd6f4",

  // Borders
  border: "#585b70",
};

/**
 * Nord Theme
 * Cool blue-grey tones with cyan accents
 * https://www.nordtheme.com/
 */
export const nord: Theme = {
  name: "nord",
  displayName: "Nord",

  // Base colors
  base: "#2e3440",
  surface: "#3b4252",
  overlay: "#434c5e",

  // Text colors
  text: "#eceff4",
  subtext: "#d8dee9",

  // Accent colors
  accent: "#88c0d0", // Frost
  accentSecondary: "#81a1c1", // Frost darker

  // Semantic colors
  success: "#a3be8c", // Aurora green
  error: "#bf616a", // Aurora red
  warning: "#ebcb8b", // Aurora yellow
  info: "#5e81ac", // Frost deep

  // Git-specific
  staged: "#8fbcbb", // Frost teal
  unstaged: "#d08770", // Aurora orange
  untracked: "#4c566a", // Polar night

  // Selection
  selection: "#434c5e",
  selectionText: "#eceff4",

  // Borders
  border: "#4c566a",
};

/**
 * Tokyo Night Theme
 * Dark blue background with vibrant purple/magenta accents
 * https://github.com/enkia/tokyo-night-vscode-theme
 */
export const tokyoNight: Theme = {
  name: "tokyo-night",
  displayName: "Tokyo Night",

  // Base colors
  base: "#1a1b26",
  surface: "#24283b",
  overlay: "#414868",

  // Text colors
  text: "#c0caf5",
  subtext: "#9aa5ce",

  // Accent colors
  accent: "#bb9af7", // Purple
  accentSecondary: "#7dcfff", // Cyan

  // Semantic colors
  success: "#9ece6a", // Green
  error: "#f7768e", // Red
  warning: "#e0af68", // Orange/Yellow
  info: "#7aa2f7", // Blue

  // Git-specific
  staged: "#73daca", // Teal
  unstaged: "#ff9e64", // Orange
  untracked: "#565f89", // Comment

  // Selection
  selection: "#364a82",
  selectionText: "#c0caf5",

  // Borders
  border: "#565f89",
};

/**
 * Dracula Theme
 * Dark purple/grey with pink, cyan, and green accents
 * https://draculatheme.com/
 */
export const dracula: Theme = {
  name: "dracula",
  displayName: "Dracula",

  // Base colors
  base: "#282a36",
  surface: "#44475a",
  overlay: "#6272a4",

  // Text colors
  text: "#f8f8f2",
  subtext: "#bfbfbf",

  // Accent colors
  accent: "#ff79c6", // Pink
  accentSecondary: "#bd93f9", // Purple

  // Semantic colors
  success: "#50fa7b", // Green
  error: "#ff5555", // Red
  warning: "#f1fa8c", // Yellow
  info: "#8be9fd", // Cyan

  // Git-specific
  staged: "#50fa7b", // Green
  unstaged: "#ffb86c", // Orange
  untracked: "#6272a4", // Comment

  // Selection
  selection: "#44475a",
  selectionText: "#f8f8f2",

  // Borders
  border: "#6272a4",
};

/**
 * All available themes
 */
export const themes: Record<ThemeName, Theme> = {
  catppuccin,
  nord,
  "tokyo-night": tokyoNight,
  dracula,
};

/**
 * Get theme by name, with fallback to catppuccin
 */
export function getTheme(name: string): Theme {
  if (name in themes) {
    return themes[name as ThemeName];
  }

  // Backward compatibility: map old values to catppuccin
  if (name === "default" || name === "dark" || name === "light") {
    return catppuccin;
  }

  return catppuccin;
}

/**
 * List of all theme names for validation
 */
export const themeNames: ThemeName[] = [
  "catppuccin",
  "nord",
  "tokyo-night",
  "dracula",
];

File: src/theme/themes.rs
use ratatui::style::Color;

/// Application theme with semantic colors
#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct Theme {
    /// Theme name
    pub name: String,
    /// Display name for the theme
    pub display_name: String,

    // Base colors
    /// Main background color
    pub base: Color,
    /// Surface/section background color
    pub surface: Color,
    /// Overlay/popup background color
    pub overlay: Color,

    // Text colors
    /// Primary text color
    pub text: Color,
    /// Secondary/dimmed text color
    pub subtext: Color,

    // Accent colors
    /// Primary accent color
    pub accent: Color,
    /// Secondary accent color
    pub accent_secondary: Color,

    // Semantic colors
    /// Success/added color (green)
    pub success: Color,
    /// Error/deleted color (red)
    pub error: Color,
    /// Warning/modified color (yellow)
    pub warning: Color,
    /// Info color (blue)
    pub info: Color,

    // Git-specific colors
    /// Staged files indicator color
    pub staged: Color,
    /// Unstaged files indicator color
    pub unstaged: Color,
    /// Untracked files color
    pub untracked: Color,

    // Selection colors
    /// Selection background color
    pub selection: Color,
    /// Selection text color
    pub selection_text: Color,

    // Border color
    pub border: Color,
}

impl Theme {
    /// Get a theme by name
    pub fn from_name(name: &str) -> Self {
        match name.to_lowercase().as_str() {
            "nord" => Self::nord(),
            "tokyo-night" | "tokyonight" => Self::tokyo_night(),
            "dracula" => Self::dracula(),
            _ => Self::catppuccin(), // Default to catppuccin
        }
    }

    /// Catppuccin Mocha theme
    pub fn catppuccin() -> Self {
        Self {
            name: "catppuccin".to_string(),
            display_name: "Catppuccin Mocha".to_string(),

            // Base colors
            base: Color::Rgb(30, 30, 46),    // #1e1e2e
            surface: Color::Rgb(49, 50, 68), // #313244
            overlay: Color::Rgb(69, 71, 90), // #45475a

            // Text colors
            text: Color::Rgb(205, 214, 244),    // #cdd6f4
            subtext: Color::Rgb(166, 173, 200), // #a6adc8

            // Accent colors
            accent: Color::Rgb(203, 166, 247), // #cba6f7 (mauve)
            accent_secondary: Color::Rgb(245, 194, 231), // #f5c2e7 (pink)

            // Semantic colors
            success: Color::Rgb(166, 227, 161), // #a6e3a1 (green)
            error: Color::Rgb(243, 139, 168),   // #f38ba8 (red)
            warning: Color::Rgb(249, 226, 175), // #f9e2af (yellow)
            info: Color::Rgb(137, 180, 250),    // #89b4fa (blue)

            // Git-specific
            staged: Color::Rgb(166, 227, 161),    // #a6e3a1 (green)
            unstaged: Color::Rgb(249, 226, 175),  // #f9e2af (yellow)
            untracked: Color::Rgb(147, 153, 178), // #9399b2 (overlay2)

            // Selection
            selection: Color::Rgb(69, 71, 90),         // #45475a
            selection_text: Color::Rgb(205, 214, 244), // #cdd6f4

            // Border
            border: Color::Rgb(88, 91, 112), // #585b70 (surface2)
        }
    }

    /// Nord theme
    pub fn nord() -> Self {
        Self {
            name: "nord".to_string(),
            display_name: "Nord".to_string(),

            // Base colors
            base: Color::Rgb(46, 52, 64),    // #2e3440 (nord0)
            surface: Color::Rgb(59, 66, 82), // #3b4252 (nord1)
            overlay: Color::Rgb(67, 76, 94), // #434c5e (nord2)

            // Text colors
            text: Color::Rgb(236, 239, 244),    // #eceff4 (nord6)
            subtext: Color::Rgb(216, 222, 233), // #d8dee9 (nord4)

            // Accent colors
            accent: Color::Rgb(136, 192, 208), // #88c0d0 (nord8 - frost)
            accent_secondary: Color::Rgb(129, 161, 193), // #81a1c1 (nord9)

            // Semantic colors
            success: Color::Rgb(163, 190, 140), // #a3be8c (nord14 - green)
            error: Color::Rgb(191, 97, 106),    // #bf616a (nord11 - red)
            warning: Color::Rgb(235, 203, 139), // #ebcb8b (nord13 - yellow)
            info: Color::Rgb(129, 161, 193),    // #81a1c1 (nord9 - blue)

            // Git-specific
            staged: Color::Rgb(163, 190, 140),   // #a3be8c (green)
            unstaged: Color::Rgb(235, 203, 139), // #ebcb8b (yellow)
            untracked: Color::Rgb(76, 86, 106),  // #4c566a (nord3)

            // Selection
            selection: Color::Rgb(67, 76, 94), // #434c5e (nord2)
            selection_text: Color::Rgb(236, 239, 244), // #eceff4

            // Border
            border: Color::Rgb(76, 86, 106), // #4c566a (nord3)
        }
    }

    /// Tokyo Night theme
    pub fn tokyo_night() -> Self {
        Self {
            name: "tokyo-night".to_string(),
            display_name: "Tokyo Night".to_string(),

            // Base colors
            base: Color::Rgb(26, 27, 38),     // #1a1b26
            surface: Color::Rgb(36, 40, 59),  // #24283b
            overlay: Color::Rgb(65, 72, 104), // #414868

            // Text colors
            text: Color::Rgb(169, 177, 214),  // #a9b1d6
            subtext: Color::Rgb(86, 95, 137), // #565f89

            // Accent colors
            accent: Color::Rgb(187, 154, 247), // #bb9af7 (purple)
            accent_secondary: Color::Rgb(255, 0, 127), // #ff007c (magenta)

            // Semantic colors
            success: Color::Rgb(158, 206, 106), // #9ece6a (green)
            error: Color::Rgb(247, 118, 142),   // #f7768e (red)
            warning: Color::Rgb(224, 175, 104), // #e0af68 (yellow)
            info: Color::Rgb(125, 207, 255),    // #7dcfff (cyan)

            // Git-specific
            staged: Color::Rgb(158, 206, 106),   // #9ece6a (green)
            unstaged: Color::Rgb(224, 175, 104), // #e0af68 (yellow)
            untracked: Color::Rgb(86, 95, 137),  // #565f89

            // Selection
            selection: Color::Rgb(65, 72, 104),        // #414868
            selection_text: Color::Rgb(169, 177, 214), // #a9b1d6

            // Border
            border: Color::Rgb(41, 46, 66), // #292e42
        }
    }

    /// Dracula theme
    pub fn dracula() -> Self {
        Self {
            name: "dracula".to_string(),
            display_name: "Dracula".to_string(),

            // Base colors
            base: Color::Rgb(40, 42, 54),      // #282a36
            surface: Color::Rgb(68, 71, 90),   // #44475a
            overlay: Color::Rgb(98, 114, 164), // #6272a4

            // Text colors
            text: Color::Rgb(248, 248, 242),   // #f8f8f2
            subtext: Color::Rgb(98, 114, 164), // #6272a4

            // Accent colors
            accent: Color::Rgb(189, 147, 249), // #bd93f9 (purple)
            accent_secondary: Color::Rgb(255, 121, 198), // #ff79c6 (pink)

            // Semantic colors
            success: Color::Rgb(80, 250, 123),  // #50fa7b (green)
            error: Color::Rgb(255, 85, 85),     // #ff5555 (red)
            warning: Color::Rgb(241, 250, 140), // #f1fa8c (yellow)
            info: Color::Rgb(139, 233, 253),    // #8be9fd (cyan)

            // Git-specific
            staged: Color::Rgb(80, 250, 123),    // #50fa7b (green)
            unstaged: Color::Rgb(241, 250, 140), // #f1fa8c (yellow)
            untracked: Color::Rgb(98, 114, 164), // #6272a4

            // Selection
            selection: Color::Rgb(68, 71, 90),         // #44475a
            selection_text: Color::Rgb(248, 248, 242), // #f8f8f2

            // Border
            border: Color::Rgb(98, 114, 164), // #6272a4
        }
    }
}

impl Default for Theme {
    fn default() -> Self {
        Self::catppuccin()
    }
}
